lib/LLVMIRCodeGen/LLVMIRGen.cpp

```diff
@@ -3180,6 +3192,107 @@ void LLVMIRGen::generateLLVMIRForInstr(llvm::IRBuilder<> &builder,
     break;
   }
 
+  case Kinded::Kind::InstrumentInstKind: {
+    auto *instrumentI = llvm::cast<InstrumentInst>(I);
+    auto *opInfo = instrumentI->getOperandsInfo();
+
+    // Instruction being instrumented.
+    Instruction *instrRef = instrumentI->getInstrRef();
+
+    // Emit instruction ID and instruction kind.
+    llvm::Type *intTy =
+        llvm::Type::getIntNTy(getLLVMContext(), getLibjitIntWidth());
+    auto *ID = llvm::ConstantInt::get(intTy, instrumentI->getID());
+    auto *kind = llvm::ConstantInt::get(intTy, (int)(instrRef->getKind()));
+
+    // Emit number of input and output operands.
+    auto inpNum = instrRef->getNumInputs();
+    auto outNum = instrRef->getNumOutputs();
+    auto opNum = inpNum + outNum;
+    auto *opInp = llvm::ConstantInt::get(intTy, inpNum);
+    auto *opOut = llvm::ConstantInt::get(intTy, outNum);
+
+    // Emit opInfo address as uint8_t*.
+    assert(opInfo->getType()->getSizeInBytes() >= 2 * sizeof(int64_t) &&
+           "Not enough memory allocated for instrumentation!");
+    auto *opInfoPtr = emitValueAddress(builder, opInfo);
+    opInfoPtr = builder.CreateBitCast(opInfoPtr, builder.getInt8PtrTy());
+
+    // Emit opAddr address as uint8_t** starting from offset 0.
+    auto *opAddrPtr =
+        builder.CreateGEP(opInfoPtr, llvm::ConstantInt::get(intTy, 0));
+    opAddrPtr = builder.CreateBitCast(opAddrPtr,
+                                      builder.getInt8PtrTy()->getPointerTo());
+
+    // Emit opSize address as int* starting from offset opNum * sizeof(int64_t).
+    auto *opSizePtr = builder.CreateGEP(
+        opInfoPtr, llvm::ConstantInt::get(intTy, opNum * sizeof(int64_t)));
+    opSizePtr = builder.CreateBitCast(opSizePtr, intTy->getPointerTo());
+
+    // Generate instrumentation.
+    auto instrumentKind = instrumentI->getInstrumentKind();
+    if (instrumentKind == InstrumentKind::Before) {
+
+      // Operands addresses and sizes.
+      std::vector<llvm::Value *> opAddrArray;
+      std::vector<llvm::Value *> opSizeArray;
+
+      // Get addresses and sizes for the input operands.
```

Comment:
Yes, the initialization occurs at run-time before the instruction. Although we could emit a constant array for the operand sizes we cannot emit a constant array for the operand addresses because those are run-time variables. The addresses depend at run-time on the base addresses provided to the bundle API. But I wouldn't worry too much because the overhead of the instrumentation is very small: for a medium sized MobileNet the run-time overhead is negligible (for an ARM Cortex-M where I tested) and the memory overhead is also negligible (the activations mem size increased with 64 bytes which is the allocation quanta of the MemoryAllocator).